core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -0,0 +1,319 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.internal.TransportType;
+
+import io.netty.channel.EventLoopGroup;
+import io.netty.resolver.AddressResolver;
+import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.HostsFileEntriesResolver;
+import io.netty.resolver.ResolvedAddressTypes;
+import io.netty.resolver.dns.DnsCnameCache;
+import io.netty.resolver.dns.DnsNameResolver;
+import io.netty.resolver.dns.DnsNameResolverBuilder;
+import io.netty.resolver.dns.DnsQueryLifecycleObserverFactory;
+import io.netty.resolver.dns.DnsServerAddressStreamProvider;
+import io.netty.resolver.dns.NoopAuthoritativeDnsServerCache;
+import io.netty.resolver.dns.NoopDnsCache;
+import io.netty.resolver.dns.NoopDnsCnameCache;
+
+/**
+ * Builds an {@link AddressResolverGroup} which builds {@link AddressResolver}s that update DNS caches
+ * automatically. Standard {@link DnsNameResolver} will only expire a cache entry after TTL,
+ * meaning DNS queries after TTL will always take time to resolve. A refreshing {@link AddressResolver}
+ * on the other hand updates the DNS cache automatically when TTL elapses,
+ * meaning DNS queries after TTL will retrieve a refreshed result right away. If refreshing fails,
+ * the {@link AddressResolver} will retry with {@link #refreshBackoff(Backoff)}.
+ *
+ * <p>The refreshing {@link AddressResolver} will only start auto refresh for a given hostname
+ * on the second access before TTL to avoid auto-refreshing for queries that only happen once
+ * (e.g., requests during server startup).
+ */
+public final class DnsResolverGroupBuilder {
+
+    private Backoff refreshBackoff = Backoff.ofDefault();
+
+    private int minTtl = 1;
+    private int maxTtl = Integer.MAX_VALUE;
+
+    // DnsNameResolverBuilder properties:
+
+    private boolean traceEnabled = true;
+    @Nullable
+    private DnsCnameCache cnameCache;
+    @Nullable
+    private Long queryTimeoutMillis;
+    @Nullable
+    private ResolvedAddressTypes resolvedAddressTypes;
+    @Nullable
+    private Boolean recursionDesired;
```

Comment:
> This flag is enabled by default.

According to the documentation of the setter of this field,`boolean recursionDesired = true`?

Reply:
All the fields are set only the user sets them. It not set, they will be set by the `DnsNameResolverBuilder` which has the default value.
Let me remove about the part explaining the default value from the Javadoc and add a link to it just in case the default value is changed and I missed that. Thanks!

Reply:
Thanks for the detail explanation! LGTM :-)